doc/design/multi_gpu_tf.md

```diff
@@ -0,0 +1,98 @@
+
+```python
+config = tf.ConfigProto()
+config.gpu_options.per_process_gpu_memory_fraction = 0.4
+session = tf.Session(config=config, ...)
+```
+
+The second requirement can be satisfied by using the `allow_growth` option. This allocates the GPU memory based on runtime allocations. When starting, it allocates very small memory, and as the Sessions get executed and if GPU memory is needed, we extend the GPU memory needed by the TensorFlow process. To turn this option on we set the option in the ConfigProto by:
+
+```python
+config = tf.ConfigProto()
+config.gpu_options.allow_growth = True
+session = tf.Session(config=config, ...)
+```
+
+This method is useful because the users can now manually control the amount of GPU memory that should be available to the TensorFlow process.
+
+## How to use a single GPU when multiple GPUs are available
+
+If our computation system has multiple GPUs, the GPU with the lowest ID is selected by default by TensorFlow. If we want to run the operation on a specific GPU, we can do so using `with tf.device()` explicitly as explained in the section above. The user has to make sure that the device mentioned exists on the system, otherwise TensorFlow throws an `InvalidArgumentError`.
+
+To avoid the above error, we can choose to allow TensorFlow to choose an existing device to run the operation in case the specified device doesn't exist (this is like the `Try` `Catch` way of handling errors). To perform this, we can set `allow_soft_placement` option to True in the configuration when creating the session as follows:
+
+```python
+sess = tf.Session(config=tf.ConfigProto(
+      allow_soft_placement=True, log_device_placement=True))
+```
+
+## Using multiple GPUs on computing devices
+
+To run a model on multiple GPUs, we can construct the model in a multi-tower fashion where each tower is assigned to a different GPU as shown in the figure below:
+<img src="./images/gpu_gpu.png" align="center"/><br/>
+
+Let us follow the following example to see how this would look like in code:
+
+```python
+c = []
+for d in ['/gpu:2', '/gpu:3']:
+  # Choose gpu:2 and gpu:3 for execution
+  with tf.device(d):
+    a = tf.constant([1.0, 2.0, 3.0, 4.0], shape=[2, 2])
+    b = tf.constant([1.0, 2.0, 3.0, 4.0], shape=[2, 2])
+    c.append(tf.matmul(a, b))
+with tf.device('/cpu:0'):
+  sum = tf.add_n(c)
+
+sess = tf.Session(config=tf.ConfigProto(log_device_placement=True))
+
+print(sess.run(sum))
+```
+Here we are running the same operation(MatMul) on two different GPUs : '/gpu:2' and '/gpu:3', and on the CPU ('/cpu:0') the operation `tf.add_n` which adds all input tensors element-wise. This operation would collect the output of MatMulfrom the GPUs and aggregate them on the CPU.
+
+The [tutorial](https://www.TensorFlow.org/tutorials/deep_cnn) on CIFAR10 is a good example demonstrating how to do training with multiple GPUs on TensorFlow.
```

Comment:
We might need more information about Device Contexts and how they are used to build new operators/kernels?

Reply:
That's a good suggestion. Will dig up more on this and add a section.